File: Algoritmos1/TPF/tipos/Ticket.hs
module Ticket (Ticket, nuevoT, salaT, peliculaT, usadoT, usarT, peliculaMenosVistaT, todosLosTicketsParaLaMismaSalaT, cambiarSalaT) 	where

import Tipos
import Pelicula

data Ticket = TicketSinUsar Sala Pelicula | TicketUsado Ticket deriving (Show, Eq)

nuevoT :: Pelicula -> Sala -> Bool -> Ticket
nuevoT p s b
	|b == False = (TicketSinUsar s p)
	|b == True = (TicketUsado (TicketSinUsar s p))
	
peliculaT :: Ticket -> Pelicula
peliculaT (TicketSinUsar _ p) = p
peliculaT (TicketUsado (TicketSinUsar _ p)) = p

salaT :: Ticket -> Sala
salaT (TicketSinUsar s _) = s
salaT (TicketUsado (TicketSinUsar s _)) = s

usadoT :: Ticket -> Bool
usadoT (TicketSinUsar _ _) = False
usadoT (TicketUsado (TicketSinUsar _ _)) = True

usarT :: Ticket -> Ticket
usarT (TicketSinUsar s p) = (TicketUsado (TicketSinUsar s p))
usarT (TicketUsado (TicketSinUsar s p)) = TicketUsado (TicketSinUsar s p)

peliculaMenosVistaT :: [Ticket] -> Pelicula
peliculaMenosVistaT xs = menosVista xs 0

cantidadVistas :: [Ticket] -> Pelicula -> Int
cantidadVistas [] p = 0
cantidadVistas (x:xs) p
	|((peliculaT x) == p) && ((usadoT x) == True) = 1 + (cantidadVistas xs p)
	|otherwise = cantidadVistas xs p

esMenosVista :: [Ticket] -> Pelicula -> Int-> Bool
esMenosVista (x:xs) p 0 = ((cantidadVistas (x:xs) p) <= (cantidadVistas (x:xs) (peliculaT x)))
esMenosVista (x:xs) p i = (cantidadVistas (x:xs) p) <= (cantidadVistas (x:xs) (peliculaT (val i (x:xs)))) && (esMenosVista (x:xs) p (i-1))

menosVista :: [Ticket] -> Int -> Pelicula
menosVista xs i
	|esMenosVista xs (peliculaT (val i xs)) ((length xs)-1) == True = peliculaT (val i xs)
	|otherwise = menosVista xs (i+1)

val :: Int -> [a] -> a
val 0 (x:xs) = x
val i (x:xs) = val (i-1) xs

todosLosTicketsParaLaMismaSalaT :: [Ticket] -> Bool
todosLosTicketsParaLaMismaSalaT [] = True
todosLosTicketsParaLaMismaSalaT [a] = True
todosLosTicketsParaLaMismaSalaT (x:y:xs) = ((salaT x) == (salaT y)) && (todosLosTicketsParaLaMismaSalaT (y:xs))

cambiarSalaT :: [Ticket] -> Sala -> Sala -> [Ticket]
cambiarSalaT [] _ _ = []
cambiarSalaT (x:xs) v n
	|(salaT x) == v && ((usadoT x) == True) = ((TicketUsado (TicketSinUsar n (peliculaT x))):(cambiarSalaT xs v n))
	|(salaT x) == v && ((usadoT x) == False) = ((TicketSinUsar n (peliculaT x)):(cambiarSalaT xs v n))
	|otherwise = (x:(cambiarSalaT xs v n))

File: Algoritmos1/TPI/codigo/ticket.cpp
#include <iostream>
#include "ticket.h"
#include <fstream>
using namespace std;

Ticket::Ticket(const Pelicula &p, Sala s, bool u){
	sala_=s;
	usado_=u;
	pelicula_=p;
}

const Pelicula& Ticket::peliculaT() const{
      return pelicula_;
}

Sala Ticket::salaT() const{
      return sala_;
}

bool Ticket::usadoT() const{
      return usado_;
}

void Ticket::usarT(){
    usado_=true;
}

void Ticket::mostrar(std::ostream& os) const{
    os<<"T ("<<pelicula_<<") "<<sala_<<" "<<usado_;
}

void Ticket::guardar(std::ostream& os) const{
    os<<"T ("<<pelicula_<<") "<<sala_<<" ";
    if(usado_ == true)
        os<<'V';
    else
        os<<'F';
}

void Ticket::cargar(std::istream& is){
    Pelicula p;
    char c;
    is>>c;
    is>>c;
    p.cargar(is);
    is>>c;
    is>>sala_;
    is>>c;
    if(c == 'V')
        usado_ = true;
    else
        usado_ = false;
    pelicula_=p;
}

std::ostream & operator<<(std::ostream & os,const Ticket & t){
    t.guardar(os);
    return os;
}

std::istream & operator>>(std::istream & is, Ticket & t){
    t.cargar(is);
    return is;
}

bool minimaCantidadDeApariciones(Lista<Pelicula> ps, Pelicula p){
    int e=0;
    while(e<ps.longitud()&&(ps.cantidadDeApariciones(ps.iesimo(e))>= ps.cantidadDeApariciones(p))){
        e++;
    }
    return e==ps.longitud();
}

Pelicula Ticket::peliculaMenosVistaT(const Lista<Ticket> &ts) const{
    Lista<Pelicula> p;
    int i=0, d=0;
    while(i<ts.longitud()){
        if(ts.iesimo(i).usado_){
            p.agregar(ts.iesimo(i).peliculaT());
        }
        i++;
    }
    i=0;
    while(i<ts.longitud()){
        if(minimaCantidadDeApariciones(p, p.iesimo(i))){
            d=i;
        }
        i++;
    }
    return p.iesimo(d);
}

bool todosIgualesInt(Lista<int> i){
    int e=0;
    bool res;
    if(i.longitud()==1){
     res=true;
    }
    else{
        while(e<i.longitud()-1 && i.iesimo(e)==i.iesimo(e+1)){
            e++;
        }
        res=(e==i.longitud()-1);
    }
    return res;
 }

bool Ticket::todosLosTicketsParaLaMismaSalaT(const Lista<Ticket> &ts) const{
    int i=0;
    Lista<int> g33;
    while(i<ts.longitud()){
        g33.agregar(ts.iesimo(i).sala_);
        i++;
    }
    return todosIgualesInt(g33);
}

Lista<Ticket> Ticket::cambiarSalaT(const Lista<Ticket> &ts, Sala vieja, Sala nueva) const{
    int i=0;
    Lista<Ticket> y;
    while(i<ts.longitud()){
        if(ts.iesimo(i).sala_==vieja){
            y.agregarAtras(Ticket(ts.iesimo(i).pelicula_, nueva, ts.iesimo(i).usado_));
        }
        else{
            y.agregarAtras(ts.iesimo(i));
        }
        i++;
    }
    return y;
}
